lightning/src/chain/mod.rs

```diff
@@ -9,6 +9,105 @@
+/// requirements.
+///
+/// Implementations **must** ensure that updates are successfully applied and persisted upon method
+/// completion. If an update fails, then it must immediately shut down without taking any further
+/// action such as persisting the current state.
+///
+/// If an implementation maintains multiple instances of a channel's monitor (e.g., by using a
+/// watchtower), then it must ensure that updates are applied across all instances. Otherwise, it
+/// could result in a revoked transaction being broadcast, allowing the counterparty to claim all
+/// funds in the channel.
+///
+/// [`ChannelMonitor`]: ../ln/channelmonitor/struct.ChannelMonitor.html
+pub trait Watch: Send + Sync {
+	/// Keys needed by monitors for creating and signing transactions.
+	type Keys: ChannelKeys;
+
+	/// Watches a channel identified by `funding_txo` using `monitor`.
+	///
+	/// Implementations are responsible for watching the chain for the funding transaction along
+	/// with any spends of outputs returned by [`get_outputs_to_watch`]. In practice, this means
+	/// calling [`block_connected`] and [`block_disconnected`] on the monitor.
+	///
+	/// [`get_outputs_to_watch`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.get_outputs_to_watch
+	/// [`block_connected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`block_disconnected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Updates a channel identified by `funding_txo` by applying `update` to its monitor.
+	///
+	/// Implementations must call [`update_monitor`] with the given update. See
+	/// [`ChannelMonitorUpdateErr`] for invariants around returning an error.
+	///
+	/// [`update_monitor`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.update_monitor
+	/// [`ChannelMonitorUpdateErr`]: ../ln/channelmonitor/enum.ChannelMonitorUpdateErr.html
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Returns any monitor events since the last call. Subsequent calls must only return new
+	/// events.
+	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent>;
+}
+
+/// The `Filter` trait defines behavior for indicating chain activity of interest pertaining to
+/// channels.
+///
+/// This is useful in order to have a [`Watch`] implementation convey to a chain source which
+/// transactions to be notified of. Notifying may take the form of pre-filtering blocks or, in the
+/// case of [BIP 157]/[BIP 158], only fetching a block if the compact filter matches.
+///
+/// [`Watch`]: trait.Watch.html
+/// [BIP 157]: https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki
+/// [BIP 158]: https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki
```

Comment:
This needs some discussion of timing requirements (ie that once the function returns, any blocks connected thereafter must contain the transaction), and a discussion about reentrancy here (also that these functions should *not* block and should queue the new monitoring data if it needs to be sent to a remote server, returning TemporaryFailure from the chain::Watch functions if you want to block, or simply rescanning if you do not).

Reply:
>  I'm not sure I follow the parts about returning TemporaryFailure from chain::Watch

I'm thinking this - lets say someone wants everything to be async (great), and they have a remote server they need to send the new data to watch. They want to return right away (cause everything is async, cool), but haven't received confirmation from the server that the data has been registered. If we're getting new data to watch from a call to block_connected, presumably they can figure out some other way to be compliant. If we're getting new data to watch from a new monitor, what they could do, instead of complying with the requirement as stated, is return `TemporaryFailure` from `watch_channel` and, once they get confirmation from the server, call `ChannelManager::channel_monitor_updated`.

Reply:
Ah, makes sense. Thanks for explaining. I've updated the documentation but left out anything about rescans until that topic has been resolved.